src/Motor.Extensions.Diagnostics.Telemetry/DistributedTracingExtensionExtensions.cs

```diff
@@ -1,20 +1,39 @@
+using System.Collections.Generic;
 using System.Diagnostics;
-using CloudNative.CloudEvents.Extensions;
+using System.Linq;
+using CloudNative.CloudEvents;
+using CloudNative.CloudEvents.Core;
+using Motor.Extensions.Hosting.CloudEvents;
 
 namespace Motor.Extensions.Diagnostics.Telemetry
 {
-    public static class DistributedTracingExtensionExtensions
+    public static class DistributedTracingExtension
     {
-        public static void SetActivity(this DistributedTracingExtension extension, Activity activity)
+        public static CloudEventAttribute TraceParentAttribute { get; } =
+            CloudEventAttribute.CreateExtension("traceparent", CloudEventAttributeType.String);
+
+        public static CloudEventAttribute TraceStateAttribute { get; } =
+            CloudEventAttribute.CreateExtension("tracestate", CloudEventAttributeType.String);
+
+        public static IEnumerable<CloudEventAttribute> AllAttributes { get; } =
+            new[] { TraceParentAttribute, TraceStateAttribute }.ToList().AsReadOnly();
+
+        public static void SetActivity<TData>(this MotorCloudEvent<TData> cloudEvent, Activity activity) where TData : class
         {
-            extension.TraceParent = activity.Id;
+            Validation.CheckNotNull(cloudEvent, nameof(cloudEvent));
+            cloudEvent[TraceParentAttribute] = activity.Id;
             if (!string.IsNullOrWhiteSpace(activity.TraceStateString))
             {
-                extension.TraceState = activity.TraceStateString;
+                cloudEvent[TraceStateAttribute] = activity.TraceStateString;
             }
         }
 
-        public static ActivityContext GetActivityContext(this DistributedTracingExtension extension) =>
-            ActivityContext.Parse(extension.TraceParent, extension.TraceState);
+        public static ActivityContext GetActivityContext<TData>(this MotorCloudEvent<TData> extension)
+            where TData : class
+        {
+            var traceParent = (string?)extension[TraceParentAttribute];
```

Comment:
+1 on the safe casts, but I personally find it a bit harder to read in this case with pattern matching. I however changed the code a bit more to this:

```c#
if (extension[TraceParentAttribute] is not string traceParent)
{
    return default;
}
var traceState = extension[TraceStateAttribute] as string;
return ActivityContext.Parse(traceParent, traceState);
```

Reply:
fair enough! I spotted another 3 places where safe-casts might be good but as already said just because I'm a huge fan of this pattern doesn't mean anyone else has to like it :smile: